File: cpp_pkg/src/turn_camera_service_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "custom_interfaces/srv/turn_camera.hpp"
#include "cv_bridge/cv_bridge.h"
#include <functional>
#include <chrono>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
class TurnCameraClient : public rclcpp::Node
{
public: 
	TurnCameraClient() : Node("turn_camera_client")
	{
		client_=this->create_client<custom_interfaces::srv::TurnCamera>("turn_camera");
		while (!client_->wait_for_service(std::chrono::seconds(1)))
		{
			if (!rclcpp::ok())
			{
				RCLCPP_ERROR(this->get_logger(), "Client interrupted while waiting for the service");
				return;
			}
			RCLCPP_INFO(this->get_logger(), "Waiting for the service to be available...");
		}
		

	}
	
	void send_request(float angle)
	{
		auto request = std::make_shared<custom_interfaces::srv::TurnCamera::Request>();
		request->degree_turn = angle;
		RCLCPP_INFO(this->get_logger(), "Sending request: angle = %f", angle);
		auto result_future = client_->async_send_request(request);

		if (rclcpp::spin_until_future_complete(this->get_node_base_interface(), result_future) ==
			rclcpp::FutureReturnCode::SUCCESS)
		{
			auto response = result_future.get();
			cv::Mat cv_image = cv_bridge::toCvCopy(
					response->camera_image, sensor_msgs::image_encodings::BGR8)->image;
			cv::imshow("Camera's View", cv_image);
			cv::waitKey(0);
		}
		else
		{
			RCLCPP_ERROR(this->get_logger(), "Failed to call service");
		}
	

	}
	private:
	
	rclcpp::Client<custom_interfaces::srv::TurnCamera>::SharedPtr client_;

};
int main(int argc, char *argv[])
{
	rclcpp::init(argc, argv);
	auto client_node = std::make_shared<TurnCameraClient>();
	float angle;
	std::cout<<"enter the camera angle: ";
	std::cin>>angle;
	client_node->send_request(angle);
	return 0;
}

File: cpp_pkg/scripts/turn_camera_service_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from custom_interfaces.srv import TurnCamera
import cv2
import cv_bridge
from sensor_msgs.msg import Image
import numpy as np


class TurnCameraClientPy(Node):
    def __init__(self):
        super().__init__('turn_camera_client_python')
        self.cli = self.create_client(TurnCamera, 'turn_camera')
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Service not available, waiting again...')
        self.req = TurnCamera.Request()
        self.bridge = cv_bridge.CvBridge()
        self.image = None
        self.angle = 0.0
        self.image_received = False

    def send_request(self):
        self.angle = input('Enter the angle to turn the camera: ')
        self.angle = float(self.angle)
        self.get_logger().info(f'Camera will turn {self.angle} degrees')
        self.req.degree_turn = self.angle
        future = self.cli.call_async(self.req)
        self.get_logger().info('Request sent')
        
        # Wait for the service response
        rclpy.spin_until_future_complete(self, future)
        
        if future.done():
            try:
                response = future.result()
                if response.camera_image is not None:
                    self.get_logger().info('Response received')
                    cv_image = self.bridge.imgmsg_to_cv2(response.camera_image, 'bgr8')
                    cv2.imshow("Camera Image", cv_image)
                    cv2.waitKey(0)
                else:
                    self.get_logger().error('Received response, but camera_image is None')
            except Exception as e:
                self.get_logger().error(f'Service call failed: {e}')
        else:
            self.get_logger().error('Service call did not complete')

def main(args=None):
    rclpy.init(args=args)
    turn_camera_client = TurnCameraClientPy()
    turn_camera_client.send_request()
    turn_camera_client.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()